state/accountsDB.go

```diff
@@ -1089,53 +1105,89 @@ func (adb *AccountsDB) SnapshotState(rootHash []byte) {
 	adb.mutOp.Lock()
 	defer adb.mutOp.Unlock()
 
-	trieStorageManager := adb.mainTrie.GetStorageManager()
-	epoch, err := trieStorageManager.GetLatestStorageEpoch()
-	if err != nil {
-		log.Error("snapshotState error", "err", err.Error())
+	trieStorageManager, epoch, shouldTakeSnapshot := adb.prepareSnapshot(rootHash)
+	if !shouldTakeSnapshot {
 		return
 	}
 
-	snapshotAlreadyTaken := bytes.Equal(adb.lastSnapshot.rootHash, rootHash) && adb.lastSnapshot.epoch == epoch
-	if !trieStorageManager.ShouldTakeSnapshot() || snapshotAlreadyTaken {
+	log.Info("starting snapshot user trie", "rootHash", rootHash, "epoch", epoch)
+	errChan := make(chan error, 1)
+	stats := newSnapshotStatistics(1)
+	go func() {
+		leavesChannel := make(chan core.KeyValueHolder, leavesChannelSize)
+		stats.NewSnapshotStarted()
+		trieStorageManager.TakeSnapshot(rootHash, rootHash, leavesChannel, errChan, stats, epoch)
+		adb.snapshotUserAccountDataTrie(true, rootHash, leavesChannel, errChan, stats, epoch)
+		trieStorageManager.ExitPruningBufferingMode()
+
+		stats.wg.Done()
+	}()
+
+	go adb.markActiveDBAfterSnapshot(stats, errChan, rootHash, "snapshotState user trie", epoch)
+
+	adb.waitForCompletionIfRunningInImportDB(stats)
+}
+
+func (adb *AccountsDB) prepareSnapshot(rootHash []byte) (common.StorageManager, uint32, bool) {
+	trieStorageManager, epoch, err := adb.getTrieStorageManagerAndLatestEpoch()
+	if err != nil {
```

Comment:
Perhaps also log the results of this call (e.g. the returned epoch?) - for debugging purposes.